gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDataset.java

```diff
@@ -35,23 +53,87 @@
+    this.props = props;
+    this.copyRoute = copyRoute;
+    this.rc = rc;
   }
 
   @Override
   public String datasetURN() {
-    // TODO Auto-generated method stub
-    return null;
+    return this.copyRoute.getCopyTo().toString();
   }
 
   @Override
   public Collection<? extends CopyEntity> getCopyableFiles(FileSystem targetFs, CopyConfiguration configuration)
       throws IOException {
-    // TODO Auto-generated method stub
-    return null;
+    List<CopyEntity> copyableFiles = Lists.newArrayList();
+    EndPoint copyFromRaw = copyRoute.getCopyFrom();
+    EndPoint copyToRaw = copyRoute.getCopyTo();
+    if (!(copyFromRaw instanceof HadoopFsEndPoint && copyToRaw instanceof HadoopFsEndPoint)) {
+      log.warn("Currently only handle the Hadoop Fs EndPoint replication");
+      return copyableFiles;
+    }
+
+    if ((!copyFromRaw.getWatermark().isPresent() && copyToRaw.getWatermark().isPresent())
+        || (copyFromRaw.getWatermark().isPresent() && copyToRaw.getWatermark().isPresent()
+            && copyFromRaw.getWatermark().get().compareTo(copyToRaw.getWatermark().get()) <= 0)) {
+      log.info(String.format(
+          "No need to copy as destination watermark >= source watermark with source watermark %s, for dataset with metadata %s",
+          copyFromRaw.getWatermark().isPresent() ? copyFromRaw.getWatermark().get().toJson() : "N/A",
+          this.rc.getMetaData()));
+      return copyableFiles;
+    }
+
+    HadoopFsEndPoint copyFrom = (HadoopFsEndPoint) copyFromRaw;
+    Configuration conf = HadoopUtils.newConfiguration();
+    FileSystem copyFromFs = FileSystem.get(copyFrom.getFsURI(), conf);
+    RecursivePathFinder finder = new RecursivePathFinder(copyFromFs, copyFrom.getDatasetPath(), this.props);
+    Set<FileStatus> copyFromFileStatuses = finder.getPaths(false);
+    
+    HadoopFsEndPoint copyTo = (HadoopFsEndPoint) copyToRaw;
+    FileSystem copyToFs = FileSystem.get(copyTo.getFsURI(), conf);
+    finder = new RecursivePathFinder(copyToFs, copyTo.getDatasetPath(), this.props);
+    Set<FileStatus> copyToFileStatuses = finder.getPaths(false);
+    Map<Path, FileStatus> copyToFileMap = new HashMap<>();
+    for (FileStatus f : copyToFileStatuses) {
+      copyToFileMap.put(PathUtils.getPathWithoutSchemeAndAuthority(f.getPath()), f);
+    }
+
+    for (FileStatus originFileStatus : copyFromFileStatuses) {
+      // construct the new path in the target file system
+      Path newPath = new Path(copyTo.getDatasetPath(),
+          PathUtils.relativizePath(originFileStatus.getPath(), copyFrom.getDatasetPath()));
+
+      if (copyToFileMap.containsKey(newPath) && copyToFileMap.get(newPath).getLen() == originFileStatus.getLen()
+          && copyToFileMap.get(newPath).getModificationTime() > originFileStatus.getModificationTime()) {
+        log.info(String.format("Skipped copy from %s for dataset with metadata %s", originFileStatus.getPath(),
```

Comment:
I'd suggest making this a `debug` level. Also it would be good to specify why it was skipped.